x-pack/plugins/canvas/public/components/var_config/edit_var.tsx

```diff
@@ -0,0 +1,177 @@
+import './var_panel.scss';
+
+interface Props {
+  selectedVar: CanvasVariable | null;
+  variables: CanvasVariable[];
+
+  onSave: (v: CanvasVariable) => void;
+  onCancel: () => void;
+}
+
+const checkDupeName = (newName: string, oldName: string | null, variables: CanvasVariable[]) => {
+  const match = variables.find((v) => {
+    // If the new name matches an existing variable and that
+    // matched variable name isn't the old name, then there
+    // is a duplicate
+    return newName === v.name && (!oldName || v.name !== oldName);
+  });
+
+  return !!match;
+};
+
+export const EditVar: FC<Props> = ({ variables, selectedVar, onCancel, onSave }) => {
+  // If there isn't a selected variable, we're creating a new var
+  const isNew = selectedVar === null;
+
+  const [type, setType] = useState(isNew ? 'string' : selectedVar!.type);
+  const [name, setName] = useState(isNew ? '' : selectedVar!.name);
+  const [value, setValue] = useState(isNew ? '' : selectedVar!.value);
+
+  const hasDupeName = checkDupeName(name, selectedVar && selectedVar.name, variables);
+
+  const typeOptions = [
+    {
+      value: 'string',
+      inputDisplay: (
+        <div className="canvasEditVar__typeOption">
+          <EuiToken iconType="tokenString" className="canvasEditVar__tokenIcon" />{' '}
+          <span>{strings.getTypeStringLabel()}</span>
+        </div>
+      ),
+    },
+    {
+      value: 'number',
+      inputDisplay: (
+        <div className="canvasEditVar__typeOption">
+          <EuiToken iconType="tokenNumber" className="canvasEditVar__tokenIcon" />{' '}
+          <span>{strings.getTypeNumberLabel()}</span>
+        </div>
+      ),
+    },
+    {
+      value: 'boolean',
+      inputDisplay: (
+        <div className="canvasEditVar__typeOption">
+          <EuiToken iconType="tokenBoolean" className="canvasEditVar__tokenIcon" />{' '}
+          <span>{strings.getTypeBooleanLabel()}</span>
+        </div>
+      ),
+    },
+  ];
+
+  return (
+    <React.Fragment>
+      <div className="canvasVarHeader__triggerWrapper">
+        <button className="canvasVarHeader__button" type="button" onClick={() => onCancel()}>
+          <span className="canvasVarHeader__iconWrapper">
+            <EuiIcon type="sortLeft" style={{ verticalAlign: 'top' }} />
+          </span>
+          <span>
+            <span className="canvasVarHeader__anchor">
+              {isNew ? strings.getAddTitle() : strings.getEditTitle()}
+            </span>
+          </span>
+        </button>
+      </div>
+      <div className="canvasSidebar__accordionContent">
+        <EuiForm component="form">
+          <EuiFormRow label={strings.getTypeFieldLabel()} display="rowCompressed">
+            <EuiSuperSelect
+              options={typeOptions}
+              valueOfSelected={type}
+              onChange={(v) => {
+                // Only have these types possible in the dropdown
+                setType(v as 'string' | 'boolean' | 'number');
```

Comment:
nit: extract as common type in `/types` somewhere, perhaps as `ValidVariableType`.  It will make it easier to find/understand, particularly as it's being cast here.

Honestly, I'd rather see this `SuperSelect` wrapped in a component which restricts the type... but it's a nitty change that is probably more trouble than it's worth.

Reply:
For the types, I changed it to use `CanvasVariable['type']`. I can pull it into it's own type still if you think it's worth it!